libfwupdplugin/fu-mem.c

```diff
@@ -941,3 +942,38 @@ fu_memwrite_uint64_safe(guint8 *buf,
 			      sizeof(tmp),
 			      error);
 }
+
+/**
+ * fu_memstrsafe:
+ * @buf: source buffer
+ * @bufsz: maximum size of @buf, typically `sizeof(buf)`
+ * @offset: offset in bytes into @buf to read from
+ * @maxsz: maximum size of returned string
+ * @error: (nullable): optional return location for an error
+ *
+ * Converts a byte buffer to a ASCII string.
+ *
+ * Returns: (transfer full): a string, or %NULL on error
+ *
+ * Since: 1.9.3
+ **/
+gchar *
+fu_memstrsafe(const guint8 *buf, gsize bufsz, gsize offset, gsize maxsz, GError **error)
+{
+	g_autofree gchar *str = NULL;
+
+	g_return_val_if_fail(buf != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
```

Comment:
do you want a check that `maxsz` isn't bigger than `bufsz`?

Reply:
I think setting the GError from fu_memchk_read() is probably better than the critical warning -- the general idea is that you can just throw any fuzzing random values into fu_memstrsafe() and get a failure rather than a critical warning.